Introduce iframe resource monitoring.

basuke opened the pull request:
#### 8b87453fe510be3928eefdb1dc1995632915525c
<pre>
Introduce iframe resource monitoring.
<a href="https://bugs.webkit.org/show_bug.cgi?id=284198">https://bugs.webkit.org/show_bug.cgi?id=284198</a>
<a href="https://rdar.apple.com/137691088">rdar://137691088</a>

Reviewed by Ben Nham.

Introduce new feature to monitor resource usage of iframe and its subresources.
The new class ResourceMonitor is the class to manage resource usage and existing
classes cooperate to send update to the class.

Monitoring is initiated by using ContentExtensionsBackend&apos;s ContentRuleList against
the urls of main frame document and subresources.

The feature is off by default.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForResourceMonitoring):
* Source/WebCore/contentextensions/ContentExtensionsBackend.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resourceMonitor):
(WebCore::Document::setResourceMonitor):
(WebCore::Document::parentResourceMonitor):
* Source/WebCore/dom/Document.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::shouldEnableResourceMonitor):
(WebCore::DocumentLoader::commitData):
* Source/WebCore/loader/ResourceLoadInfo.h:
(WebCore::ContentExtensions::ResourceLoadInfo::isolatedCopy const):
(WebCore::ContentExtensions::ResourceLoadInfo::isolatedCopy):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::didReceiveResponse):
(WebCore::ResourceLoader::didReceiveBuffer):
(WebCore::ResourceLoader::resourceMonitor):
* Source/WebCore/loader/ResourceLoader.h:
* Source/WebCore/loader/ResourceMonitor.cpp: Added.
(WebCore::networkUsageThresholdWithRandomNoise):
(WebCore::ResourceMonitor::create):
(WebCore::ResourceMonitor::ResourceMonitor):
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::didReceiveResponse):
(WebCore::ResourceMonitor::addNetworkUsage):
(WebCore::ResourceMonitor::checkNetworkUsageExcessIfNecessary):
(WebCore::ResourceMonitor::protectedFrame const):
(WebCore::ResourceMonitor::parentResourceMonitor const):
* Source/WebCore/loader/ResourceMonitor.h: Added.
* Source/WebCore/loader/ResourceMonitorChecker.cpp: Added.
(WebCore::ResourceMonitorChecker::singleton):
(WebCore::ResourceMonitorChecker::ResourceMonitorChecker):
(WebCore::ResourceMonitorChecker::checkEligibility):
(WebCore::ResourceMonitorChecker::checkEligibilityForTesting):
(WebCore::ResourceMonitorChecker::checkEligibilityWithLock):
(WebCore::ResourceMonitorChecker::setContentRuleList):
* Source/WebCore/loader/ResourceMonitorChecker.h: Added.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::networkUsageDidExceedThreshold):
(WebCore::LocalFrame::showResourceMonitoringError):
* Source/WebCore/page/LocalFrame.h:

Canonical link: <a href="https://commits.webkit.org/287667@main">https://commits.webkit.org/287667@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74f34f3fc8a2e5e29882143ff1a806dc70163f23

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80328 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59334 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84849 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31310 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82439 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68396 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7635 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62793 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20606 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83397 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52867 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73153 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43096 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27308 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29769 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73308 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71325 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27823 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86283 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79387 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7553 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5349 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71074 "Found 5 new test failures: compositing/animation/keyframe-order.html fast/text/crash-letter-spacing-infinite.html imported/w3c/web-platform-tests/css/css-transforms/animation/translate-percent-with-width-and-height-separate.html webanimations/accelerated-animation-addition-lower-in-effect-stack.html webanimations/accelerated-animation-easing-update-steps-after-pause.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7728 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68991 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70314 "Found 1 new API test failure: /TestWebKit:WebKit.PageLoadState (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17534 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14307 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13260 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101794 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7515 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13035 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24808 "Found 3 new JSC stress test failures: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-collect-continuously, wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-eager, wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-eager-jettison (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7354 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10874 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9160 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1381,6 +1383,59 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
+
+ResourceMonitor* LocalFrame::resourceMonitor()
+{
+    constexpr bool forceEnableResourceMonitoring = true;
```
Why is this here? It looks like debugging code.

cdumez commented on Source/WebCore/loader/ResourceMonitor.h:
```diff
@@ -0,0 +1,99 @@
+    using ResourceLoadInfo = ContentExtensions::ResourceLoadInfo;
+    using QueryCallback = Function<void(Eligibility)>;
+
+    class Checker final {
```
Why is this public?

basuke replied:
I'm going to add API test for Checker so that it is better to be public than adding friend. Any suggestion?

cdumez replied:
You can make it be its own class in its own file.

cdumez commented on Source/WebCore/loader/ResourceMonitor.h:
```diff
@@ -0,0 +1,99 @@
+    WeakRef<LocalFrame> m_frame;
+    URL m_frameURL;
+    Eligibility m_eligibility { Eligibility::Unsure };
+    bool m_networkUsageExceed { };
```
`{ false }`

basuke commented on Source/WebCore/loader/ResourceMonitor.h:
```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2015 Apple Inc. All rights reserved.
```
Copyright year.

cdumez commented on Source/WebCore/loader/ResourceLoader.cpp:
```diff
@@ -601,6 +602,11 @@ void ResourceLoader::didReceiveResponse(const ResourceResponse& r, CompletionHan
+
+#if ENABLE(CONTENT_EXTENSIONS)
+    if (RefPtr monitor = resourceMonitor())
+        monitor->didReceiveResource(m_response.url(), m_resourceType);
```
I would call this `didReceiveResponse` for consistency.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+
+#if ENABLE(CONTENT_EXTENSIONS)
+
+static constexpr size_t sNetworkUsageThreshold = 4 * 1024 * 1024; // 4MB
```
`networkUsageThreshold`, no prefixes for static variables.

cdumez replied:
You can use `4 * MB` is you include `<wtf/StdLibExtras.h>

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+#if ENABLE(CONTENT_EXTENSIONS)
+
+static constexpr size_t sNetworkUsageThreshold = 4 * 1024 * 1024; // 4MB
+static constexpr size_t sNetworkUsageThresholdRandomness = 1303 * 1024; // 1303KB
```
Ditto.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+        .type = resourceType
+    };
+
+    Checker::singleton().checkEligibility(info, [weakThis = WeakPtr(this)](Eligibility eligibility) mutable {
```
We usually use `WeakPtr { *this }`

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+
+void ResourceMonitor::addNetworkUsage(size_t bytes)
+{
+    m_networkUsage += bytes;
```
It seems we should guard against overflows here, maybe using CheckedArithmetics.h.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+        checkCondition();
+}
+
+void ResourceMonitor::checkCondition()
```
`checkCondition()` is too generic a name. Maybe `checkNetworkUsageExcessIfNecessary`

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+    if (m_eligibility != Eligibility::Eligible)
+        return;
+
+    if (!m_networkUsageExceed && m_networkUsage >= 4 * 1024 * 1024) {
```
Could use `* MB`

basuke replied:
`m_networkUsageThreshold` should be used here.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+
+ResourceMonitor::ResourceMonitor(LocalFrame& frame)
+    : m_frame(frame)
+    , m_networkUsageThreshold { networkUsageThresholdWithRandomNoise() }
```
What is this used for? I don't see it used in this PR. We should not land dead code.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+{
+}
+
+static ResourceMonitor::ResourceLoadInfo isolatedCopy(const ResourceMonitor::ResourceLoadInfo& info)
```
This should be a member function on `ResourceMonitor::ResourceLoadInfo` like we normally do.

basuke replied:
There's so simple to be extracted as a member function so embedding this in `checkEligibility`.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+            eligibility = checkEligibilityWithLock(info);
+        }
+
+        callback(eligibility);
```
We should not be calling the callback on another thread. This is a foot-gun and is bound to cause thread safety bugs. We should dispatch back to the main thread before calling the callback.

cdumez commented on Source/WebCore/loader/ResourceMonitor.h:
```diff
@@ -0,0 +1,99 @@
+    public:
+        static Checker& singleton();
+
+        Checker();
```
The constructor should NOT be public.

cdumez commented on Source/WebCore/loader/ResourceMonitor.h:
```diff
@@ -0,0 +1,99 @@
+    Eligibility eligibility() const { return m_eligibility; }
+    void setEligibility(Eligibility);
+
+    void setDocument(URL&&);
```
`setDocumentURL`

cdumez commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1381,6 +1383,59 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
+    for (RefPtr<Frame> frame = this; frame; frame = frame->tree().traverseNext()) {
+        if (RefPtr localFrame = dynamicDowncast<LocalFrame>(frame)) {
+            if (RefPtr window = localFrame->window())
+                window->removeAllEventListeners();
```
I don't think we should do this if the `RefPtr iframeElement = dynamicDowncast<HTMLIFrameElement>(ownerElement())` check fails later on. We'd break content without offloading it.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+
+void ResourceMonitor::Checker::checkEligibility(const ResourceLoadInfo& info, QueryCallback&& callback)
+{
+    m_workQueue->dispatch([this, info = isolatedCopy(info), callback = WTFMove(callback)] {
```
Since this class uses multiple thread, please add `ASSERT(isMainThread());` or `ASSERT(!isMainThread());` at the beginning of every member function to document what thread they're meant to be called on.

cdumez commented on Source/WebCore/loader/ResourceMonitor.h:
```diff
@@ -0,0 +1,99 @@
+    enum class Eligibility : uint8_t { Unsure, NotEligible, Eligible };
+    using ResourceTypeSet = OptionSet<ContentExtensions::ResourceType>;
+    using ResourceLoadInfo = ContentExtensions::ResourceLoadInfo;
+    using QueryCallback = Function<void(Eligibility)>;
```
This should be a `CompletionHandler` since it is called only once and needs to be called at least once.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -0,0 +1,185 @@
+static constexpr size_t sNetworkUsageThreshold = 4 * 1024 * 1024; // 4MB
+static constexpr size_t sNetworkUsageThresholdRandomness = 1303 * 1024; // 1303KB
+
+static size_t networkUsageThresholdWithRandomNoise()
```
Where is this used?

cdumez commented on Source/WebCore/page/LocalFrame.h:
```diff
@@ -404,6 +415,10 @@ class LocalFrame final : public Frame {
     HashSet<RegistrableDomain> m_storageAccessExceptionDomains;
+
+#if ENABLE(CONTENT_EXTENSIONS)
+    RefPtr<ResourceMonitor> m_resourceMonitor;
```
Do frames ever share ResourceMonitors? Shouldn't same-origin iframes share the same monitor? Otherwise, what's preventing a frame from creating many same-origin subframes to do a lot of networking within them?

basuke replied:
I think about that design but are there any chance the frame is moved into different frame? In that case, I cannot think of easy way to recalculate the usage. Or even we need to recalculate the usage or not. Then it's safe to make the monitor for each iframe.

cdumez replied:
Can we store on the Document then? Documents definitely do not get moved. Frames can change URL (be navigated) and I don't think we want to track resources usage across navigations anyway?

basuke replied:
That sounds great. I try that.

basuke replied:
Let me clarify your suggestion. You suggested that the `ResourceMonitor` should be shared among same origin frames and to make the network usage won't be affected when iframe is moved into different parent, store the usage in `document` instead of in `ResourceMonitor` directly. Is this correct?

cdumez replied:
No, I'm asking if the `ResourceMonitor` should be on `Document` instead of `Frame`. A Frame gets re-used on navigation, a Document doesn't.

basuke replied:
> A Frame gets re-used on navigation

Oh, didn't know that. Got it.

basuke replied:
I may misunderstand the initial comment. I thought you've asked that two same origin frames should share one `ResourceMonitor`. Is this something you've concerned or not?

cdumez replied:
Yes, I do think that Documents sharing the same origin should share the same ResourceMonitor or at least the same quota. Otherwise, it is too easy to work around limits by creating new same-origin iframes. It doesn't have to be in this first PR but we'll have to deal with it before enabling the feature.

cdumez reviewed (commented):
Doesn't have to be dealt with in this PR what we're going to have to account for networking done by workers too. Otherwise it is trivial to spin a worker, have it do networking and then postMessage back the data. I could be wrong but I don't think this is currently being handled (which is fine as a first cut).

bnham commented on Source/WebCore/loader/ResourceLoader.cpp:
```diff
@@ -601,6 +602,11 @@ void ResourceLoader::didReceiveResponse(const ResourceResponse& r, CompletionHan
+
+#if ENABLE(CONTENT_EXTENSIONS)
+    if (RefPtr monitor = resourceMonitor())
+        monitor->didReceiveResponce(m_response.url(), m_resourceType);
```
nit: `Response` not `Responce`

bnham commented on Source/WebCore/loader/ResourceMonitorChecker.cpp:
```diff
@@ -0,0 +1,99 @@
+        .mainFrameContext = info.mainFrameContext,
+    };
+
+    protectedWorkQueue()->dispatch([this, info = WTFMove(info), completionHandler = WTFMove(completionHandler)] mutable {
```
I think the usual pattern for something like this would be to define `isolatedCopy() const` and `isolatedCopy() &&` on `ContentExtensions::ResourceLoadInfo`. Then when you want to pass that object across threads, do `info = crossThreadCopy(info)` or `info = crossThreadCopy(WTFMove(info))`.

cdumez replied:
Correct. info needs an isolatedCopy() function and we should call `info = crossThreadCopy(WTFMove(info))` here.

bnham commented on Source/WebCore/loader/ResourceMonitorChecker.cpp:
```diff
@@ -0,0 +1,99 @@
+
+ResourceMonitor::Eligibility ResourceMonitorChecker::checkEligibilityWithLock(const ContentExtensions::ResourceLoadInfo& info)
+{
+    if (!m_ruleList)
+        return Eligibility::Unsure;
```
It seems like if you have an `Unsure` state then you will have to reschedule the eligibility check some time later after the content rule list is set.

If we are pretty sure that the content rule list will be set in a timely manner, maybe it makes more sense to get rid of the lock and the `Unsure` state? For instance, use `SuspendableWorkQueue` instead of `WorkQueue`. In the constructor, suspend the work queue. Then resume the work queue in the first call to `setContentRuleList`.

This way, every function that executes in the work queue can know that it's executing with a non-null `m_ruleList`. This also lets us get rid of the lock.

basuke replied:
Nice. I've never heard that convenient class. I'll use it.

basuke replied:
I will switch implementation in next patch where actual rule list is delivered.

bnham commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1381,6 +1383,40 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
+    if (RefPtr protectedWindow = window(); protectedWindow && protectedWindow->hasStickyActivation())
+        return;
+
+    showResourceMonitoringError("This frame was unloaded."_s);
```
We should `RELEASE_LOG` something here so we know this happened when examining system logs.

cdumez commented on Source/WebCore/loader/ResourceMonitorChecker.cpp:
```diff
@@ -0,0 +1,91 @@
+
+ResourceMonitorChecker& ResourceMonitorChecker::singleton()
+{
+    static NeverDestroyed<ResourceMonitorChecker> globalChecker;
```
`MainThreadNeverDestroyed` or if this can be accessed across multiple threads, use `LazyNeverDestroyed + std::call_once()`

cdumez commented on Source/WebCore/loader/ResourceMonitorChecker.h:
```diff
@@ -0,0 +1,68 @@
+public:
+    using Eligibility = ResourceMonitor::Eligibility;
+
+    ResourceMonitorChecker();
```
This should not be public.

cdumez commented on Source/WebCore/loader/ResourceMonitorChecker.h:
```diff
@@ -0,0 +1,68 @@
+
+class ResourceMonitorChecker final {
+public:
+    friend class ResourceMonitor;
```
No need.

cdumez commented on Source/WebCore/loader/ResourceMonitorChecker.h:
```diff
@@ -0,0 +1,68 @@
+    void setContentRuleList(ContentExtensions::ContentExtensionsBackend&&);
+
+private:
+    static ResourceMonitorChecker& singleton();
```
This should be public.

cdumez commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1381,6 +1383,42 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
+    if (RefPtr protectedWindow = window(); protectedWindow && protectedWindow->hasStickyActivation())
+        return;
+
+    FRAME_RELEASE_LOG_ERROR(ResourceLoading, "networkUsageDidExceedThreshold: Unloading frame by exceeding threshold.");
```
"by" -> "due to"

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -11076,6 +11077,28 @@ void Document::securityOriginDidChange()
+    m_resourceMonitor = WTFMove(resourceMonitor);
+}
+
+ResourceMonitor* Document::parentResourceMonitor()
```
Don't we want the top-most ResourceMonitor? What if there is frame nesting like so?
```
A
  B - Eligible
    C
      D - Shouldn't this look up B's monitor?
```

basuke replied:
When frame is created inside eligible frame, that frame is also eligible in the constructor and no check is done further more.

bnham replied:
Ok, I talked to Basuke about this, and he points out that B, C, and D should always have a resource monitor (only the root doesn't have one). When you call `ResourceMonitor::addNetworkUsage` on D, it reports network usage up the tree all the way up to A unconditionally.

basuke replied:
And resource monitor is always created except for the main frame and no chance if C doesn't have resource monitor.

basuke committed:
Introduce iframe resource monitoring.
https://bugs.webkit.org/show_bug.cgi?id=284198
rdar://137691088

Reviewed by Ben Nham.

Introduce new feature to monitor resource usage of iframe and its subresources.
The new class ResourceMonitor is the class to manage resource usage and existing
classes cooperate to send update to the class.

Monitoring is initiated by using ContentExtensionsBackend's ContentRuleList against
the urls of main frame document and subresources.

The feature is off by default.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForResourceMonitoring):
* Source/WebCore/contentextensions/ContentExtensionsBackend.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resourceMonitor):
(WebCore::Document::setResourceMonitor):
(WebCore::Document::parentResourceMonitor):
* Source/WebCore/dom/Document.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::shouldEnableResourceMonitor):
(WebCore::DocumentLoader::commitData):
* Source/WebCore/loader/ResourceLoadInfo.h:
(WebCore::ContentExtensions::ResourceLoadInfo::isolatedCopy const):
(WebCore::ContentExtensions::ResourceLoadInfo::isolatedCopy):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::didReceiveResponse):
(WebCore::ResourceLoader::didReceiveBuffer):
(WebCore::ResourceLoader::resourceMonitor):
* Source/WebCore/loader/ResourceLoader.h:
* Source/WebCore/loader/ResourceMonitor.cpp: Added.
(WebCore::networkUsageThresholdWithRandomNoise):
(WebCore::ResourceMonitor::create):
(WebCore::ResourceMonitor::ResourceMonitor):
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::didReceiveResponse):
(WebCore::ResourceMonitor::addNetworkUsage):
(WebCore::ResourceMonitor::checkNetworkUsageExcessIfNecessary):
(WebCore::ResourceMonitor::protectedFrame const):
(WebCore::ResourceMonitor::parentResourceMonitor const):
* Source/WebCore/loader/ResourceMonitor.h: Added.
* Source/WebCore/loader/ResourceMonitorChecker.cpp: Added.
(WebCore::ResourceMonitorChecker::singleton):
(WebCore::ResourceMonitorChecker::ResourceMonitorChecker):
(WebCore::ResourceMonitorChecker::checkEligibility):
(WebCore::ResourceMonitorChecker::checkEligibilityForTesting):
(WebCore::ResourceMonitorChecker::checkEligibilityWithLock):
(WebCore::ResourceMonitorChecker::setContentRuleList):
* Source/WebCore/loader/ResourceMonitorChecker.h: Added.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::networkUsageDidExceedThreshold):
(WebCore::LocalFrame::showResourceMonitoringError):
* Source/WebCore/page/LocalFrame.h:

Canonical link: https://commits.webkit.org/287667@main
